runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -763,12 +763,19 @@ public final void snapshotState(StateSnapshotContext context) throws Exception {
 
     // We can't output here anymore because the checkpoint barrier has already been
     // sent downstream. This is going to change with 1.6/1.7's prepareSnapshotBarrier.
-    outputManager.openBuffer();
-    // Ensure that no new bundle gets started as part of finishing a bundle
-    while (bundleStarted.get()) {
-      invokeFinishBundle();
+    try {
+      outputManager.openBuffer();
+      // Ensure that no new bundle gets started as part of finishing a bundle
+      while (bundleStarted.get()) {
+        invokeFinishBundle();
+      }
+      outputManager.closeBuffer();
+    } catch (Exception e) {
```

Comment:
Maybe also create a JIRA in Flink to have a cleaner way to signal an application error?

Reply:
Would make sense. Normally, there should be no application logic in here, though there are probably other people who have similar logic in `snapshotState`.

Reply:
https://jira.apache.org/jira/browse/FLINK-14653